### PR TITLE
feat: backport automod

### PR DESCRIPTION
**Pull request by jaw0r3k**
**Please describe the changes this PR makes and why it should be merged:**

Backports #7938
Backports #8954
Backports #8962
Backports #8977

API types are added to rawTypes file

**Status and versioning classification:**

- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: src/util/Constants.js [hunk 1]
```diff
@@ -1219,14 +1229,44 @@ exports.ApplicationCommandOptionTypes = createEnum([
...
+ */
+exports.AutoModerationActionTypes = createEnum([null, 'BLOCK_MESSAGE', 'SEND_ALERT_MESSAGE', 'TIMEOUT']);
+
+/**
+ * The type of an {@link AutoModerationRuleEventTypes} object:
+ * * MESSAGE_SEND
+ * @typedef {string} AutoModerationRuleEventType
+ * @see {@link https://discord.com/developers/docs/resources/auto-moderation#auto-moderation-rule-object-event-types}
  */
-exports.ApplicationCommandPermissionTypes = createEnum([null, 'ROLE', 'USER']);
```
Comment (Syjalo) [inline src/util/Constants.js:1228]: Why did you remove it?

File: src/managers/AutoModerationRuleManager.js [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+  _add(data, cache) {
+    return super._add(data, cache, { extras: [this.guild] });
+  }
+
+  /**
+   * Options used to set the trigger metadata of an auto moderation rule.
+   * @typedef {Object} AutoModerationTriggerMetadataOptions
+   * @property {string[]} [keywordFilter] The substrings that will be searched for in the content
+   * @property {(string|RegExp)[]} [regexPatterns] The regular expression patterns
```
Comment (almeidx) [inline src/managers/AutoModerationRuleManager.js:36]: This should not accept RegExp instances
https://github.com/discordjs/discord.js/pull/7938#discussion_r1020837942

File: src/structures/AutoModerationRule.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+   * @param {string} [reason] The reason for changing the keyword filter of this auto moderation rule
+   * @returns {Promise<AutoModerationRule>}
+   */
+  setKeywordFilter(keywordFilter, reason) {
+    return this.edit({ triggerMetadata: { ...this.triggerMetadata, keywordFilter }, reason });
+  }
+
+  /**
+   * Sets the regular expression patterns for this auto moderation rule.
+   * @param {(string|RegExp)[]} regexPatterns The regular expression patterns of this auto moderation rule
```
Comment (almeidx) [inline src/structures/AutoModerationRule.js:198]: This should not accept RegExp instances
https://github.com/discordjs/discord.js/pull/7938#discussion_r1020837942

File: typings/index.d.ts [hunk 4]
```diff
@@ -4195,7 +4334,13 @@ export interface ClientEvents extends BaseClientEvents {
   /** @deprecated See [this issue](https://github.com/discord/discord-api-docs/issues/3690) for more information. */
   applicationCommandDelete: [command: ApplicationCommand];
   /** @deprecated See [this issue](https://github.com/discord/discord-api-docs/issues/3690) for more information. */
-  applicationCommandUpdate: [oldCommand: ApplicationCommand | null, newCommand: ApplicationCommand];
```
Comment (Syjalo) [inline typings/index.d.ts:4198]: ditto

**Review by Syjalo — CHANGES_REQUESTED** [review]

**Review by almeidx — CHANGES_REQUESTED** [review]

File: src/util/Constants.js (hunk 1 above)
Comment (jaw0r3k) [reply src/util/Constants.js:1228]: idk missclick

File: typings/rawDataTypes.d.ts [hunk 5]
```diff
@@ -216,3 +223,48 @@ export type RawWelcomeScreenData = APIGuildWelcomeScreen;
 export type RawWidgetData = APIGuildWidget;
 
 export type RawWidgetMemberData = APIGuildWidgetMember;
+
+export interface GatewayAutoModerationActionExecutionDispatchData {
```
Comment (vladfrangu) [inline typings/rawDataTypes.d.ts:227]: Couldn't these be imported from discord-api-types?

File: src/managers/AutoModerationRuleManager.js [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+  _add(data, cache) {
+    return super._add(data, cache, { extras: [this.guild] });
+  }
+
+  /**
+   * Options used to set the trigger metadata of an auto moderation rule.
+   * @typedef {Object} AutoModerationTriggerMetadataOptions
+   * @property {string[]} [keywordFilter] The substrings that will be searched for in the content
+   * @property {string[]} [regexPatterns] The regular expression patterns
+   *  which will be matched against the content
```
Comment (kyranet) [inline src/managers/AutoModerationRuleManager.js:37]:
```suggestion
   * @property {string[]} [regexPatterns] The regular expression patterns which will be matched against the content
```

File: src/managers/AutoModerationRuleManager.js [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+        trigger_metadata: triggerMetadata && {
+          keyword_filter: triggerMetadata.keywordFilter,
+          regex_patterns: triggerMetadata.regexPatterns,
+          presets: triggerMetadata.presets?.map(preset =>
+            typeof preset === 'number' ? preset : AutoModerationRuleKeywordPresetTypes[preset],
+          ),
+          allow_list: triggerMetadata.allowList,
+          mention_total_limit: triggerMetadata.mentionTotalLimit,
+        },
+        actions: actions?.map(action => ({
```
Comment (kyranet) [inline src/managers/AutoModerationRuleManager.js:112]:
```suggestion
        actions: actions.map(action => ({
```
v14: https://github.com/discordjs/discord.js/blob/3160ab0fc75946041cbe7a093032ea10503dfd8c/packages/discord.js/src/managers/AutoModerationRuleManager.js#L128

File: src/structures/AutoModerationActionExecution.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * The id of any system auto moderation messages posted as a result of this action.
+     * @type {?Snowflake}
+     */
+    this.alertSystemMessageId = data.alert_system_message_id ?? null;
+
+    /**
+     * The content that triggered this action.
+     * <info>This property requires the {@link GatewayIntentBits.MessageContent} privileged gateway intent.</info>
```
Comment (kyranet) [inline src/structures/AutoModerationActionExecution.js:61]:
```suggestion
     * <info>This property requires the {@link Intents.FLAGS.MESSAGE_CONTENT} privileged gateway intent.</info>
```

File: src/structures/AutoModerationRule.js [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+      /**
+       * Additional data used to determine whether an auto moderation rule should be triggered.
+       * @typedef {Object} AutoModerationTriggerMetadata
+       * @property {string[]} keywordFilter The substrings that will be searched for in the content
+       * @property {string[]} regexPatterns The regular expression patterns which will be matched against the content
+       * <info>Only Rust-flavored regular expressions are supported.</info>
+       * @property {AutoModerationRuleKeywordPresetType[]} presets
+       * The internally pre-defined wordsets which will be searched for in the content
+       * @property {string[]} allowList The substrings that will be exempt from triggering
+       * {@link AutoModerationRuleTriggerTypes.Keyword} and {@link AutoModerationRuleTriggerTypes.KeywordPreset}
```
Comment (kyranet) [inline src/structures/AutoModerationRule.js:74]:
```suggestion
       * {@link AutoModerationRuleTriggerTypes.KEYWORD} and {@link AutoModerationRuleTriggerTypes.KEYWORD_PRESET}
```

File: src/structures/AutoModerationRule.js [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  /**
+   * Sets the keyword filter for this auto moderation rule.
+   * @param {string[]} keywordFilter The keyword filter of this auto moderation rule
+   * @param {string} [reason] The reason for changing the keyword filter of this auto moderation rule
+   * @returns {Promise<AutoModerationRule>}
+   */
+  setKeywordFilter(keywordFilter, reason) {
+    return this.edit({ triggerMetadata: { ...this.triggerMetadata, keywordFilter }, reason });
```
Comment (kyranet) [inline src/structures/AutoModerationRule.js:193]:
```suggestion
    return this.edit({ triggerMetadata: { keywordFilter }, reason });
```
v14: https://github.com/discordjs/discord.js/blob/3160ab0fc75946041cbe7a093032ea10503dfd8c/packages/discord.js/src/structures/AutoModerationRule.js#L187

File: src/structures/AutoModerationRule.js [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  /**
+   * Sets the presets for this auto moderation rule.
+   * @param {AutoModerationRuleKeywordPresetType[]} presets The presets of this auto moderation rule
+   * @param {string} [reason] The reason for changing the presets of this auto moderation rule
+   * @returns {Promise<AutoModerationRule>}
+   */
+  setPresets(presets, reason) {
+    return this.edit({ triggerMetadata: { ...this.triggerMetadata, presets }, reason });
```
Comment (kyranet) [inline src/structures/AutoModerationRule.js:214]:
```suggestion
    return this.edit({ triggerMetadata: { presets }, reason });
```

v14: https://github.com/discordjs/discord.js/blob/3160ab0fc75946041cbe7a093032ea10503dfd8c/packages/discord.js/src/structures/AutoModerationRule.js#L208

File: src/structures/AutoModerationRule.js [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+  /**
+   * Sets the regular expression patterns for this auto moderation rule.
+   * @param {string[]} regexPatterns The regular expression patterns of this auto moderation rule
+   * <info>Only Rust-flavored regular expressions are supported.</info>
+   * @param {string} [reason] The reason for changing the regular expression patterns of this auto moderation rule
+   * @returns {Promise<AutoModerationRule>}
+   */
+  setRegexPatterns(regexPatterns, reason) {
+    return this.edit({ triggerMetadata: { ...this.triggerMetadata, regexPatterns }, reason });
```
Comment (kyranet) [inline src/structures/AutoModerationRule.js:204]:
```suggestion
    return this.edit({ triggerMetadata: { regexPatterns }, reason });
```
v14: https://github.com/discordjs/discord.js/blob/3160ab0fc75946041cbe7a093032ea10503dfd8c/packages/discord.js/src/structures/AutoModerationRule.js#L198

File: src/structures/AutoModerationRule.js [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  /**
+   * Sets the allow list for this auto moderation rule.
+   * @param {string[]} allowList The allow list of this auto moderation rule
+   * @param {string} [reason] The reason for changing the allow list of this auto moderation rule
+   * @returns {Promise<AutoModerationRule>}
+   */
+  setAllowList(allowList, reason) {
+    return this.edit({ triggerMetadata: { ...this.triggerMetadata, allowList }, reason });
```
Comment (kyranet) [inline src/structures/AutoModerationRule.js:224]:
```suggestion
    return this.edit({ triggerMetadata: { allowList }, reason });
```
v14: https://github.com/discordjs/discord.js/blob/3160ab0fc75946041cbe7a093032ea10503dfd8c/packages/discord.js/src/structures/AutoModerationRule.js#L218

File: src/structures/AutoModerationRule.js [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  /**
+   * Sets the mention total limit for this auto moderation rule.
+   * @param {number} mentionTotalLimit The mention total limit of this auto moderation rule
+   * @param {string} [reason] The reason for changing the mention total limit of this auto moderation rule
+   * @returns {Promise<AutoModerationRule>}
+   */
+  setMentionTotalLimit(mentionTotalLimit, reason) {
+    return this.edit({ triggerMetadata: { ...this.triggerMetadata, mentionTotalLimit }, reason });
```
Comment (kyranet) [inline src/structures/AutoModerationRule.js:234]:
```suggestion
    return this.edit({ triggerMetadata: { mentionTotalLimit }, reason });
```
v14: https://github.com/discordjs/discord.js/blob/3160ab0fc75946041cbe7a093032ea10503dfd8c/packages/discord.js/src/structures/AutoModerationRule.js#L228

**Review by kyranet — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: src/structures/AutoModerationRule.js (hunk 14 above)
Comment (jaw0r3k) [reply src/structures/AutoModerationRule.js:234]: then we need to change it in v14 
as it wont save previous settings

File: src/structures/AutoModerationRule.js (hunk 13 above)
Comment (jaw0r3k) [reply src/structures/AutoModerationRule.js:224]: the same 
> then we need to change it in v14 
> as it wont save previous settings

File: src/structures/AutoModerationRule.js (hunk 11 above)
Comment (jaw0r3k) [reply src/structures/AutoModerationRule.js:214]: the same 
> then we need to change it in v14 
> as it wont save previous settings

File: src/structures/AutoModerationRule.js (hunk 12 above)
Comment (jaw0r3k) [reply src/structures/AutoModerationRule.js:204]: the same 
> then we need to change it in v14 
> as it wont save previous settings

File: src/structures/AutoModerationRule.js (hunk 10 above)
Comment (jaw0r3k) [reply src/structures/AutoModerationRule.js:193]: > the same 
> then we need to change it in v14 
> as it wont save previous settings

**Review by kyranet — APPROVED** [review]

File: src/structures/AutoModerationRule.js (hunk 14 above)
Comment (kyranet) [reply src/structures/AutoModerationRule.js:234]: @Jiralite confirmed this, good to know, and thank you for opening a PR to fix this bug in v14. 😓

**Review by Jiralite — CHANGES_REQUESTED** [review]
Needs changes from:

- #8954
- #8962

File: src/managers/AutoModerationRuleManager.js [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+   */
+
+  /**
+   * Resolves an {@link AutoModerationRuleResolvable} to a {@link AutoModerationRule} id.
+   * @method resolveId
+   * @memberof AutoModerationRuleManager
+   * @instance
+   * @param {AutoModerationRuleResolvable} autoModerationRule The AutoModerationRule resolvable to resolve
+   * @returns {?Snowflake}
+   */
```
Comment (kyranet) [inline src/managers/AutoModerationRuleManager.js:285]: Those need to be moved above, see #8954.

File: src/structures/AutoModerationRule.js [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  /**
+   * Sets the actions for this auto moderation rule.
+   * @param {AutoModerationActionOptions} actions The actions of this auto moderation rule
```
Comment (kyranet) [inline src/structures/AutoModerationRule.js:239]:
```suggestion
   * @param {AutoModerationActionOptions[]} actions The actions of this auto moderation rule
```

See #8962.

File: typings/index.d.ts [hunk 17]
```diff
@@ -3989,6 +4003,84 @@ export type ApplicationCommandType = keyof typeof ApplicationCommandTypes;
...
+  public setEventType(
+    eventType: AutoModerationRuleEventType | AutoModerationRuleEventTypes,
+    reason?: string,
+  ): Promise<AutoModerationRule>;
+  public setKeywordFilter(keywordFilter: string[], reason?: string): Promise<AutoModerationRule>;
+  public setRegexPatterns(regexPatterns: string[], reason?: string): Promise<AutoModerationRule>;
+  public setPresets(presets: AutoModerationRuleKeywordPresetType[], reason?: string): Promise<AutoModerationRule>;
+  public setAllowList(allowList: string[], reason?: string): Promise<AutoModerationRule>;
+  public setMentionTotalLimit(mentionTotalLimit: number, reason?: string): Promise<AutoModerationRule>;
+  public setActions(actions: AutoModerationActionOptions, reason?: string): Promise<AutoModerationRule>;
```
Comment (kyranet) [inline typings/index.d.ts:4055]:
```suggestion
  public setActions(actions: AutoModerationActionOptions[], reason?: string): Promise<AutoModerationRule>;
```

See #8962.

File: typings/index.d.ts [hunk 18]
```diff
@@ -4013,6 +4105,57 @@ export type ApplicationFlagsString =
...
+export interface FetchAutoModerationRulesOptions {
+  cache?: boolean;
+}
+
+export interface AutoModerationRuleCreateOptions {
+  name: string;
+  eventType: AutoModerationRuleEventType | AutoModerationRuleEventTypes;
+  triggerType: AutoModerationRuleTriggerType | AutoModerationRuleTriggerTypes;
+  triggerMetadata?: AutoModerationTriggerMetadataOptions;
+  actions: AutoModerationActionOptions;
```
Comment (kyranet) [inline typings/index.d.ts:4139]:
```suggestion
  actions: AutoModerationActionOptions[];
```

See #8962.

**Review by kyranet — CHANGES_REQUESTED** [review]
Missed a couple of backports, thanks @Jiralite for reporting them internally!

File: src/util/Constants.js [hunk 19]
```diff
@@ -1411,6 +1460,11 @@ function createEnum(keys) {
  * The type of an {@link ApplicationCommandPermissions} object.
  * @property {Object<ApplicationCommandType, number>} ApplicationCommandTypes
  * The type of an {@link ApplicationCommand} object.
+ * @property {Object<AutoModerationRuleTriggerType, number>} AutoModerationRuleTriggerTypes Characterizes the type
+ * of contentwhich can trigger the rule.
+ * @property {Object<AutoModerationActionType, number>} AutoModerationActionTypes
+ * @property {Object<AutoModerationRuleKeywordPresetType, number>} AutoModerationRuleKeywordPresetTypes
+ * @property {Object<AutoModerationRuleEventType, number>} AutoModerationRuleEventTypes
```
Comment (Jiralite) [inline src/util/Constants.js:1467]: Typos & missing descriptions here

**Review by Jiralite — CHANGES_REQUESTED** [review]

File: src/managers/AutoModerationRuleManager.js [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+  constructor(guild, iterable) {
+    super(guild.client, AutoModerationRule, iterable);
+
+    /**
+     * The guild this manager belongs to.
+     * @type {Guild}
+     */
+    this.guild = guild;
+  }
+  /**
```
Comment (kyranet) [inline src/managers/AutoModerationRuleManager.js:27]:
```suggestion
  }

  /**
```

**Review by kyranet — CHANGES_REQUESTED** [review]
One last thing

**Review by kyranet — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Review by Jiralite — CHANGES_REQUESTED** [review]
`MessageTypes` is missing `AUTO_MODERATION_ACTION`.

**Review by Jiralite — CHANGES_REQUESTED** [review]
This is also missing the `AUTO_MODERATION` guild feature.
